File: src/main/java/com/learnspark/notification/scheduler/ReminderScheduler.java
package com.learnspark.notification.scheduler;

import com.learnspark.notification.config.ReminderProperties;
import com.learnspark.notification.service.ReminderService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

/**
 * 提醒定时任务：每分钟扫描到点用户并发送邮件。
 *
 * <p>cron 表达式由 app.reminder.cron 配置，默认 "0 * * * * ?"（每分钟整点）。
 */
@Slf4j
@Component
@RequiredArgsConstructor
public class ReminderScheduler {

    private final ReminderService reminderService;
    private final ReminderProperties reminderProperties;

    /**
     * 每分钟扫描并发送提醒。
     *
     * <p>注意：cron 通过属性配置注入，避免硬编码。
     */
    @Scheduled(cron = "${app.reminder.cron:0 * * * * ?}")
    public void scanAndSend() {
        try {
            int sent = reminderService.scanAndSendReminders();
            if (sent > 0) {
                log.info("定时提醒任务完成: 发送 {} 封邮件", sent);
            }
        } catch (Exception e) {
            log.error("定时提醒任务异常: {}", e.getMessage(), e);
        }
    }
}

File: learnspark-kmp/server/src/main/kotlin/com/learnspark/server/service/ReminderScheduler.kt
package com.learnspark.server.service

import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component

/**
 * R3b：每分钟跑一次的调度器，扫描到期提醒并落库日志。
 *
 * 设计要点：
 * - 进程内单实例（app-api 跑，app-worker 也可同时跑 —— 必须保证幂等性）
 *   通过 [ReminderService.tick] 在事务内一次性读 + 写，最坏情况：双 worker 各跑一次，触发两条日志
 *   → 生产可加 SELECT ... FOR UPDATE SKIP LOCKED（MySQL 8+）保证独占；本阶段接受幂等重复
 * - 失败容错：tick 内部 try/catch 已在 Service 层做，本组件不抛
 *
 * 调度频率：fixedDelay = 60s（任务结束 60s 后再启动，避免与上一次重叠）
 */
@Component
class ReminderScheduler(
    private val reminderService: ReminderService,
) {
    private val log = LoggerFactory.getLogger(ReminderScheduler::class.java)

    @Scheduled(fixedDelay = 60_000L, initialDelay = 15_000L)
    fun run() {
        try {
            val r = reminderService.tick()
            if (r.fired > 0) {
                log.info("ReminderScheduler tick: fired={} rescheduled={}", r.fired, r.rescheduled)
            }
        } catch (e: Exception) {
            // 永远不让 scheduler 自己挂掉下一轮
            log.warn("ReminderScheduler tick failed: {}", e.message)
        }
    }
}
